File: client/flutter_chat/lib/models/enums.dart
/// Chat scope. Serialized as "Private" / "Group" to match the server
/// (server uses [JsonConverter(JsonStringEnumConverter)]).
enum ChatType { private, group }

/// Message kind. Serialized as "Text" / "Image" / "File" / "Voice" to match the server.
enum MessageType { text, image, file, voice }

ChatType chatTypeFromJson(String? v) {
  switch (v) {
    case 'Group':
      return ChatType.group;
    case 'Private':
    default:
      return ChatType.private;
  }
}

String chatTypeToJson(ChatType t) => t == ChatType.group ? 'Group' : 'Private';

MessageType messageTypeFromJson(String? v) {
  switch (v) {
    case 'Image':
      return MessageType.image;
    case 'File':
      return MessageType.file;
    case 'Voice':
      return MessageType.voice;
    case 'Text':
    default:
      return MessageType.text;
  }
}

String messageTypeToJson(MessageType t) {
  switch (t) {
    case MessageType.image:
      return 'Image';
    case MessageType.file:
      return 'File';
    case MessageType.voice:
      return 'Voice';
    case MessageType.text:
      return 'Text';
  }
}

/// Matches server CallType (JsonStringEnumConverter serializes as "Voice" / "Video").
enum CallType { voice, video }

CallType callTypeFromJson(String? v) {
  switch (v) {
    case 'Video':
      return CallType.video;
    case 'Voice':
    default:
      return CallType.voice;
  }
}

String callTypeToJson(CallType t) => t == CallType.video ? 'Video' : 'Voice';

/// Matches server CallState.
enum CallState { calling, connecting, connected, ended }

CallState callStateFromJson(String? v) {
  switch (v) {
    case 'Connecting':
      return CallState.connecting;
    case 'Connected':
      return CallState.connected;
    case 'Ended':
      return CallState.ended;
    case 'Calling':
    default:
      return CallState.calling;
  }
}

/// Matches server CallEndReason.
enum CallEndReason { declined, busy, timeout, hangUp, offline, error }

CallEndReason callEndReasonFromJson(String? v) {
  switch (v) {
    case 'Declined':
      return CallEndReason.declined;
    case 'Busy':
      return CallEndReason.busy;
    case 'Timeout':
      return CallEndReason.timeout;
    case 'HangUp':
      return CallEndReason.hangUp;
    case 'Offline':
      return CallEndReason.offline;
    case 'Error':
    default:
      return CallEndReason.error;
  }
}

String callEndReasonKey(CallEndReason r) {
  switch (r) {
    case CallEndReason.declined:
      return '对方已拒绝';
    case CallEndReason.busy:
      return '对方忙线';
    case CallEndReason.timeout:
      return '对方无应答';
    case CallEndReason.hangUp:
      return '通话结束';
    case CallEndReason.offline:
      return '对方不在线';
    case CallEndReason.error:
      return '连接失败';
  }
}

File: server/Chat.Shared/Entities/Enums.cs
using System.Text.Json.Serialization;

namespace Chat.Shared.Entities;

// [JsonConverter] 註解保證：無論走 AddJsonProtocol 的全局選項還是屬性級反射，
// "Text"/"Image"/"File" 這類字符串都能可靠地與 enum 互轉，避免客戶端默認值與
// 服務端枚舉不匹配時報告 "Error binding arguments"。
[JsonConverter(typeof(JsonStringEnumConverter))]
public enum ChatType
{
    Private = 0,
    Group = 1
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum MessageType
{
    Text = 0,
    Image = 1,
    File = 2,
    Voice = 3
}

public enum FriendshipStatus
{
    Pending = 0,
    Accepted = 1
}

public enum GroupMemberRole
{
    Member = 0,
    Admin = 1,
    Owner = 2
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum CallType
{
    Voice = 0,
    Video = 1
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum CallState
{
    Calling = 0,
    Connecting = 1,
    Connected = 2,
    Ended = 3
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum CallEndReason
{
    Declined = 0,
    Busy = 1,
    Timeout = 2,
    HangUp = 3,
    Offline = 4,
    Error = 5
}
